Structure panel

YanisBft opened the pull request:
This pull request adds a structure panel to Enigma.

Structure panel is loaded when a class is opened or switched, and refreshed when a member is renamed.
The structure tree shows inner classes, methods and fields of a class, with fancy new icons (shoutout to @Juuxel). These icons are also displayed in obf/deobf panels.
The panel also has an option to hide deobfuscated members.

It has been tested in many different scenarios, including local multiplayer.

Closes #213 
Closes #302 

Video preview:
https://streamable.com/dgmv8m

YTG1234 reviewed (approved):
A few suggestions that can be implemented after this PR is merged:
* Different icons for obfuscated/deobfuscated members
* Different icons for client-only and dedicated server exclusive members

haykam821 commented:
I think that these should have icon variants (while keeping the same general color):

- Interfaces
- Enums
- Constructors
- Static members

Gegy commented on enigma/src/main/java/cuchaz/enigma/analysis/StructureTreeNode.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    private boolean isDeobfuscated(EnigmaProject project, ParentedEntry child) {
+        List<ObfuscationTestService> obfuscationTestServices = project.getEnigma().getServices().get(ObfuscationTestService.TYPE);
```
It isn't very important- but is this similar logic replicated anywhere else? It might be nice to be able to deduplicate it since this logic doesn't seem unique to here.

YanisBft replied:
There was the same logic in StatsGenerator, moved it to EnigmaProject

Juuxel commented:
@haykam821 At least in the obf/deobf panels that would need some refactoring. Currently, those tree nodes mainly store info about the class names and packages, but not other properties like whether it's an interface.

Constructors would be easy to add though since they're just `<init>`.

2xsaiko commented on enigma/src/main/java/cuchaz/enigma/analysis/index/JarIndex.java:
```diff
@@ -40,6 +37,7 @@
 	private final Collection<JarIndexer> indexers;
 
 	private final Multimap<String, MethodDefEntry> methodImplementations = HashMultimap.create();
+	private final Map<ClassEntry, List<ParentedEntry>> childrenByClass;
```
Since Multimap is a thing, it might actually be better to use that here since it's effectively the same but less annoying to use.

YanisBft replied:
Multimap has an important drawback, entries are not sorted anymore by indexing order (inner classes, then fields, then methods)

Juuxel replied:
There's a `ListMultimap` that uses `List` as the value type, which should work here.

YanisBft replied:
Changed to ListMultimap, thanks Juuz

2xsaiko reviewed (changes requested):
Couple things I noticed:

- When you double click an item in the structure panel, it jumps to that entry. However, there's no selection box on the list if you click an item, and if you double click a class which has subentries, it jumps to the class definition but also collapses the children (and there's no expander icon or anything to show that the children can be expanded again, at least on the top level class).
- Double clicking an inner class entry opens that inner class in a new tab instead of the outer class's tab; not sure if this is intended but it can't be done any other way afaik.
- Double clicking with middle mouse and right mouse button also jumps to the entry, which is a bit weird, you should probably check for the left mouse button explicitly.

Otherwise this is looking pretty good!

A thing that might be worth exploring (although not in this PR) is to never construct new *Entry instances except at jar load, and just looking them up from an existing collection when needed. That way you could store the children entries directly in the object since there's only one object for each entry, and it would probably also reduce memory usage by at least a bit.

YanisBft commented:
> When you double click an item in the structure panel, it jumps to that entry. However, there's no selection box on the list if you click an item, and if you double click a class which has subentries, it jumps to the class definition but also collapses the children (and there's no expander icon or anything to show that the children can be expanded again, at least on the top level class).

The selection box is now fixed, it was due to the custom tree cell renderer. Regarding the second part, seems to be related to your other concern below?

> Double clicking an inner class entry opens that inner class in a new tab instead of the outer class's tab; not sure if this is intended but it can't be done any other way afaik.

Yeah this is a global issue in Enigma, also happens when double clicking an inner class in other panels, and when ctrl+clicking an inner class token in the editor panel.

> Double clicking with middle mouse and right mouse button also jumps to the entry, which is a bit weird, you should probably check for the left mouse button explicitly.

Fixed for the structure panel, as well as inheritance, implementation and calls panels.

2xsaiko commented:
The top level entry still has no visible toggle. To be honest I don't know how easy it is to add that, if it's more involved than just setting a flag somewhere just leave it. I'd just add it because it could be confusing if you collapse it accidentally without noticing and then wondering why the class has no members (and also because there's no way to collapse/expand without double clicking which also opens the class again). But this is what I mean:
![image](https://user-images.githubusercontent.com/3987560/102700719-6688f280-4250-11eb-9e16-a2cf74002c1a.png)
vs 
![image](https://user-images.githubusercontent.com/3987560/102700726-76083b80-4250-11eb-800e-b7e27392c430.png)

YanisBft commented:
Ah, didn't even the issue as it's not happening with dark theme for some reasons. Now fixed, for all the panels

2xsaiko reviewed (approved):
Thanks!
